src/snowflake/snowpark/_internal/server_connection.py

```diff
@@ -183,6 +185,12 @@ def __init__(
             "_skip_upload_on_content_match" in signature.parameters
         )
 
+    @property
+    def _cursor(self) -> SnowflakeCursor:
+        if not hasattr(self._thread_stored, "cursor"):
+            self._thread_stored.cursor = self._conn.cursor()
+        return self._thread_stored.cursor
```

Comment:
does this operation need to be protected by self._lock?

Reply:
this is a `threading.local()` object so each thread has its own copy. We don't need a lock for this step.